File: java/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonTriangulation/monotoneDecomposition/_TriangulationTrapezoid.java
package vsdk.toolkit.environment.geometry.geometricProcessing.polygonTriangulation.monotoneDecomposition;

final class _TriangulationTrapezoid {
    // [SEID1991].3 Bounding segment indices. The original reference allocates
    // the trapezoid table with calloc, so unassigned bounding segments default
    // to 0, which is the reserved zeroed "no segment" slot (segmentAt(0)). All
    // logic compares these fields with > 0 / <= 0, so 0 behaves as "absent"
    // while keeping segmentAt() in bounds for degenerate trapezoids (a negative
    // sentinel would index out of the segment table and crash).
    int leftSegmentIndex = 0;
    int rightSegmentIndex = 0;
    _Point2D upperPoint = new _Point2D();
    _Point2D lowerPoint = new _Point2D();
    int upperLeftTrapezoidIndex;
    int upperRightTrapezoidIndex;
    int lowerLeftTrapezoidIndex;
    int lowerRightTrapezoidIndex;
    _TriangulationTrapezoidQueryNode sinkNode;
    int savedUpperNeighborIndex;
    int savedUpperNeighborSide;
    int status = 1;

    void copyFrom(_TriangulationTrapezoid other) {
        leftSegmentIndex = other.leftSegmentIndex;
        rightSegmentIndex = other.rightSegmentIndex;
        upperPoint.set(other.upperPoint);
        lowerPoint.set(other.lowerPoint);
        upperLeftTrapezoidIndex = other.upperLeftTrapezoidIndex;
        upperRightTrapezoidIndex = other.upperRightTrapezoidIndex;
        lowerLeftTrapezoidIndex = other.lowerLeftTrapezoidIndex;
        lowerRightTrapezoidIndex = other.lowerRightTrapezoidIndex;
        sinkNode = other.sinkNode;
        savedUpperNeighborIndex = other.savedUpperNeighborIndex;
        savedUpperNeighborSide = other.savedUpperNeighborSide;
        status = other.status;
    }

    int insidePolygon() {
        int rightBoundarySegmentIndex = this.rightSegmentIndex;

        if (status == _Construct.ST_INVALID) return 0;
        if ((leftSegmentIndex <= 0) || (rightBoundarySegmentIndex <= 0)) return 0;
        if (((upperLeftTrapezoidIndex <= 0) && (upperRightTrapezoidIndex <= 0)) ||
            ((lowerLeftTrapezoidIndex <= 0) && (lowerRightTrapezoidIndex <= 0))) {
            return _Construct.greaterThan(
                _Construct.segmentAt(rightBoundarySegmentIndex).endPoint,
                _Construct.segmentAt(rightBoundarySegmentIndex).startPoint) ? 1 : 0;
        }
        return 0;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonTriangulation/monotoneDecomposition/_TriangulationTrapezoid.ts
import { Vector2Dd } from "../../../../../common/linealAlgebra/Vector2Dd.js";
import { _Construct } from "./_Construct.js";
import type { _TriangulationTrapezoidQueryNode } from "./_TriangulationTrapezoidQueryNode.js";
export class _TriangulationTrapezoid {
    public leftSegmentIndex = 0;
    public rightSegmentIndex = 0;
    public upperPoint = new Vector2Dd();
    public lowerPoint = new Vector2Dd();
    public upperLeftTrapezoidIndex = 0;
    public upperRightTrapezoidIndex = 0;
    public lowerLeftTrapezoidIndex = 0;
    public lowerRightTrapezoidIndex = 0;
    public sinkNode: _TriangulationTrapezoidQueryNode | null = null;
    public savedUpperNeighborIndex = 0;
    public savedUpperNeighborSide = 0;
    public status = _Construct.ST_VALID;
    public copyFrom(x: _TriangulationTrapezoid): void {
        Object.assign(this, x);
        this.upperPoint = new Vector2Dd(x.upperPoint);
        this.lowerPoint = new Vector2Dd(x.lowerPoint);
    }
    public insidePolygon(): number {
        if (this.status === _Construct.ST_INVALID || this.leftSegmentIndex <= 0 || this.rightSegmentIndex <= 0)
            return 0;
        if (
            (this.upperLeftTrapezoidIndex <= 0 && this.upperRightTrapezoidIndex <= 0) ||
            (this.lowerLeftTrapezoidIndex <= 0 && this.lowerRightTrapezoidIndex <= 0)
        )
            return _Construct.greaterThan(
                _Construct.segmentAt(this.rightSegmentIndex).endPoint,
                _Construct.segmentAt(this.rightSegmentIndex).startPoint,
            )
                ? 1
                : 0;
        return 0;
    }
}
